File: src/main/java/toby/helpers/Cache.java
package toby.helpers;

import org.apache.commons.collections4.MapIterator;
import org.apache.commons.collections4.map.LRUMap;

import java.util.ArrayList;

public class Cache <K, T> {

    private final long timeToLiveInMillis;

    private final LRUMap<K, CachedObject> cacheMap;

    protected class CachedObject {
        public long lastAccessed = System.currentTimeMillis();
        public T value;

        protected CachedObject(T value) {
            this.value = value;
        }
    }

    public Cache(long timeToLiveInSeconds, final long timerIntervalInSeconds,
                 int maxItems) {
        this.timeToLiveInMillis = timeToLiveInSeconds * 1000;

        cacheMap = new LRUMap<>(maxItems);

        if (timeToLiveInMillis > 0 && timerIntervalInSeconds > 0) {

            Thread t = new Thread(() -> {
                while (true) {
                    try {
                        Thread.sleep(timerIntervalInSeconds * 1000);
                    }
                    catch (InterruptedException ignored) {
                    }

                    cleanup();
                }
            });

            t.setDaemon(true);
            t.start();
        }
    }

    public void put(K key, T value) {
        synchronized (cacheMap) {
            cacheMap.put(key, new CachedObject(value));
        }
    }

    public T get(K key) {
        synchronized (cacheMap) {
            CachedObject c = cacheMap.get(key);

            if (c == null)
                return null;
            else {
                c.lastAccessed = System.currentTimeMillis();
                return c.value;
            }
        }
    }

    public void remove(K key) {
        synchronized (cacheMap) {
            cacheMap.remove(key);
        }
    }

    public int size() {
        synchronized (cacheMap) {
            return cacheMap.size();
        }
    }

    public void cleanup() {

        long now = System.currentTimeMillis();
        ArrayList<K> keysToDelete;

        synchronized (cacheMap) {
            MapIterator<K, CachedObject> itr = cacheMap.mapIterator();

            keysToDelete = new ArrayList<>((cacheMap.size() / 2) + 1);
            K key;
            CachedObject c;

            while (itr.hasNext()) {
                key = itr.next();
                c = itr.getValue();

                if (c != null && (now > (timeToLiveInMillis + c.lastAccessed))) {
                    keysToDelete.add(key);
                }
            }
        }

        for (K key : keysToDelete) {
            synchronized (cacheMap) {
                cacheMap.remove(key);
            }

            Thread.yield();
        }
    }
}

File: discord-bot/src/main/kotlin/bot/toby/helpers/Cache.kt
package bot.toby.helpers

import org.apache.commons.collections4.MapIterator
import org.apache.commons.collections4.map.LRUMap

open class Cache(
    timeToLiveInSeconds: Long, timerIntervalInSeconds: Long,
    maxItems: Int
) {
    private val timeToLiveInMillis = timeToLiveInSeconds * 1000

    private val cacheMap: LRUMap<String, CachedObject> = LRUMap(maxItems)

    protected inner class CachedObject(var value: List<String>) {
        var lastAccessed: Long = System.currentTimeMillis()
    }

    init {

        if (timeToLiveInMillis > 0 && timerIntervalInSeconds > 0) {
            val t = Thread {
                while (true) {
                    try {
                        Thread.sleep(timerIntervalInSeconds * 1000)
                    } catch (ignored: InterruptedException) {
                    }

                    cleanup()
                }
            }

            t.isDaemon = true
            t.start()
        }
    }

    fun put(key: String, value: List<String>) {
        synchronized(cacheMap) {
            cacheMap.put(key, CachedObject(value))
        }
    }

    fun get(key: String): List<String>? {
        synchronized(cacheMap) {
            val c = cacheMap[key]
            if (c == null) return null
            else {
                c.lastAccessed = System.currentTimeMillis()
                return c.value
            }
        }
    }

    fun remove(key: String) {
        synchronized(cacheMap) {
            cacheMap.remove(key)
        }
    }

    fun size(): Int {
        synchronized(cacheMap) {
            return cacheMap.size
        }
    }

    private fun cleanup() {
        val now = System.currentTimeMillis()
        var keysToDelete: ArrayList<String>

        synchronized(cacheMap) {
            val itr: MapIterator<String, CachedObject> = cacheMap.mapIterator()
            keysToDelete = ArrayList((cacheMap.size / 2) + 1)
            var key: String
            var c: CachedObject?
            while (itr.hasNext()) {
                key = itr.next()
                c = itr.value

                if (c != null && (now > (timeToLiveInMillis + c.lastAccessed))) {
                    keysToDelete.add(key)
                }
            }
        }

        for (key in keysToDelete) {
            synchronized(cacheMap) {
                cacheMap.remove(key)
            }

            Thread.yield()
        }
    }
}
